File: src/domain/models/gap_analysis.py
"""GapAnalysis aggregate root and supporting value objects.

Manages the lifecycle of a rescue operation (alty init --existing) through
a strict state machine: SCANNING -> ANALYZED -> PLANNED -> EXECUTING ->
COMPLETED (or FAILED from EXECUTING).

The core invariant is **scan-before-analyze**: no analysis without a scan,
no plan without analysis, no execution without a plan.
"""

from __future__ import annotations

import enum
import uuid
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pathlib import Path

    from src.domain.events.rescue_events import GapAnalysisCompleted

from src.domain.models.errors import InvariantViolationError as InvariantViolationError


class GapType(enum.Enum):
    """Classification of a structural gap found during project analysis.

    MISSING_DOC:       A required documentation file is absent.
    MISSING_CONFIG:    A required configuration file is absent.
    MISSING_STRUCTURE: A required directory structure is absent.
    MISSING_TOOLING:   A required tooling config is absent.
    MISSING_KNOWLEDGE: The knowledge base directory is absent.
    CONFLICT:          An existing file conflicts with the expected structure.
    """

    MISSING_DOC = "missing_doc"
    MISSING_CONFIG = "missing_config"
    MISSING_STRUCTURE = "missing_structure"
    MISSING_TOOLING = "missing_tooling"
    MISSING_KNOWLEDGE = "missing_knowledge"
    CONFLICT = "conflict"


@dataclass(frozen=True)
class ProjectScan:
    """Immutable snapshot of an existing project's current state.

    Attributes:
        project_dir: Root directory of the scanned project.
        existing_docs: Tuple of documentation file paths found.
        existing_configs: Tuple of configuration file paths found.
        existing_structure: Tuple of directory structure paths found.
        has_knowledge_dir: Whether .alty/knowledge/ exists.
        has_agents_md: Whether AGENTS.md exists at the project root.
        has_git: Whether the project is a git repository.
        has_tests: Whether a tests/ directory exists.
    """

    project_dir: Path
    existing_docs: tuple[str, ...] = ()
    existing_configs: tuple[str, ...] = ()
    existing_structure: tuple[str, ...] = ()
    has_knowledge_dir: bool = False
    has_agents_md: bool = False
    has_git: bool = False
    has_tests: bool = False


@dataclass(frozen=True)
class Gap:
    """A single structural gap found during project analysis.

    Attributes:
        gap_id: Unique identifier for this gap.
        gap_type: Classification of the gap.
        path: Relative path where the gap was detected.
        description: Human-readable explanation of the gap.
        severity: How critical the gap is: "required", "recommended", or "optional".
    """

    gap_id: str
    gap_type: GapType
    path: str
    description: str
    severity: str  # "required" | "recommended" | "optional"


@dataclass(frozen=True)
class MigrationPlan:
    """Immutable plan for resolving gaps in an existing project.

    Attributes:
        plan_id: Unique identifier for this plan.
        gaps: Tuple of gaps to resolve.
        branch_name: Git branch to create for the migration.
        skip_agents_md: Whether to skip generating AGENTS.md (already exists).
    """

    plan_id: str
    gaps: tuple[Gap, ...]
    branch_name: str = "alty/init"
    skip_agents_md: bool = False


class AnalysisStatus(enum.Enum):
    """States in the gap analysis lifecycle.

    SCANNING:   Initial state, project is being scanned.
    ANALYZED:   Gaps have been identified.
    PLANNED:    A migration plan has been created.
    EXECUTING:  The plan is being executed.
    COMPLETED:  Execution finished successfully.
    FAILED:     Execution encountered an error.
    """

    SCANNING = "scanning"
    ANALYZED = "analyzed"
    PLANNED = "planned"
    EXECUTING = "executing"
    COMPLETED = "completed"
    FAILED = "failed"


class GapAnalysis:
    """Aggregate root for the rescue flow.

    Enforces the scan-before-analyze invariant and produces domain events
    on completion.

    Attributes:
        analysis_id: Unique identifier for this analysis.
        project_dir: The directory being analyzed.
    """

    def __init__(self, project_dir: Path) -> None:
        self.analysis_id: str = str(uuid.uuid4())
        self.project_dir: Path = project_dir
        self._status: AnalysisStatus = AnalysisStatus.SCANNING
        self._scan: ProjectScan | None = None
        self._gaps: tuple[Gap, ...] = ()
        self._plan: MigrationPlan | None = None
        self._events: list[GapAnalysisCompleted] = []

    @property
    def status(self) -> AnalysisStatus:
        """Current analysis state."""
        return self._status

    @property
    def scan(self) -> ProjectScan | None:
        """The project scan result, or None if not yet scanned."""
        return self._scan

    @property
    def gaps(self) -> tuple[Gap, ...]:
        """Identified structural gaps."""
        return self._gaps

    @property
    def plan(self) -> MigrationPlan | None:
        """The migration plan, or None if not yet planned."""
        return self._plan

    @property
    def events(self) -> list[GapAnalysisCompleted]:
        """Domain events produced by this aggregate (defensive copy)."""
        return list(self._events)

    def set_scan(self, scan: ProjectScan) -> None:
        """Record scan results. Only from SCANNING state.

        Args:
            scan: The project scan result.

        Raises:
            InvariantViolationError: If not in SCANNING state.
        """
        if self._status != AnalysisStatus.SCANNING:
            msg = f"Cannot set scan in {self._status.value} state"
            raise InvariantViolationError(msg)
        self._scan = scan

    def analyze(self, gaps: tuple[Gap, ...]) -> None:
        """Set analysis results. Requires scan to exist.

        Args:
            gaps: The identified gaps.

        Raises:
            InvariantViolationError: If not in SCANNING state or scan is missing.
        """
        if self._status != AnalysisStatus.SCANNING:
            msg = f"Cannot analyze in {self._status.value} state"
            raise InvariantViolationError(msg)
        if self._scan is None:
            raise InvariantViolationError("Cannot analyze without scan")
        self._gaps = gaps
        self._status = AnalysisStatus.ANALYZED

    def create_plan(self, plan: MigrationPlan) -> None:
        """Create migration plan from gaps. Only from ANALYZED.

        Args:
            plan: The migration plan to apply.

        Raises:
            InvariantViolationError: If not in ANALYZED state.
        """
        if self._status != AnalysisStatus.ANALYZED:
            msg = f"Cannot create plan in {self._status.value} state"
            raise InvariantViolationError(msg)
        self._plan = plan
        self._status = AnalysisStatus.PLANNED

    def begin_execution(self) -> None:
        """Start executing the plan. Only from PLANNED.

        Raises:
            InvariantViolationError: If not in PLANNED state.
        """
        if self._status != AnalysisStatus.PLANNED:
            msg = f"Cannot begin execution in {self._status.value} state"
            raise InvariantViolationError(msg)
        self._status = AnalysisStatus.EXECUTING

    def complete(self) -> None:
        """Mark as completed and emit event.

        Raises:
            InvariantViolationError: If not in EXECUTING state.
        """
        from src.domain.events.rescue_events import GapAnalysisCompleted

        if self._status != AnalysisStatus.EXECUTING:
            msg = f"Cannot complete in {self._status.value} state"
            raise InvariantViolationError(msg)
        self._status = AnalysisStatus.COMPLETED
        self._events.append(
            GapAnalysisCompleted(
                analysis_id=self.analysis_id,
                project_dir=self.project_dir,
                gaps_found=len(self._gaps),
                gaps_resolved=len(self._gaps),
            )
        )

    def fail(self, reason: str) -> None:
        """Mark as failed. Only from EXECUTING.

        Args:
            reason: Human-readable failure reason.

        Raises:
            InvariantViolationError: If not in EXECUTING state.
        """
        if self._status != AnalysisStatus.EXECUTING:
            msg = f"Cannot fail in {self._status.value} state"
            raise InvariantViolationError(msg)
        self._status = AnalysisStatus.FAILED

File: internal/rescue/domain/gap_analysis.go
package domain

import (
	"fmt"
	"strings"

	domainerrors "github.com/alty-cli/alty/internal/shared/domain/errors"
	"github.com/alty-cli/alty/internal/shared/domain/identity"
)

// GapType classifies a structural gap found during project analysis.
type GapType string

// Gap type constants.
const (
	GapTypeMissingDoc       GapType = "missing_doc"
	GapTypeMissingConfig    GapType = "missing_config"
	GapTypeMissingStructure GapType = "missing_structure"
	GapTypeMissingTooling   GapType = "missing_tooling"
	GapTypeMissingKnowledge GapType = "missing_knowledge"
	GapTypeConflict         GapType = "conflict"
)

// GapSeverity indicates how critical a gap is.
type GapSeverity string

// Gap severity constants.
const (
	GapSeverityRequired    GapSeverity = "required"
	GapSeverityRecommended GapSeverity = "recommended"
	GapSeverityOptional    GapSeverity = "optional"
)

// AllGapSeverities returns all gap severity values.
func AllGapSeverities() []GapSeverity {
	return []GapSeverity{GapSeverityRequired, GapSeverityRecommended, GapSeverityOptional}
}

// AnalysisStatus represents states in the gap analysis lifecycle.
type AnalysisStatus string

// Analysis status constants.
const (
	AnalysisStatusScanning  AnalysisStatus = "scanning"
	AnalysisStatusAnalyzed  AnalysisStatus = "analyzed"
	AnalysisStatusPlanned   AnalysisStatus = "planned"
	AnalysisStatusExecuting AnalysisStatus = "executing"
	AnalysisStatusCompleted AnalysisStatus = "completed"
	AnalysisStatusFailed    AnalysisStatus = "failed"
)

// ProjectScan is an immutable snapshot of an existing project's current state.
type ProjectScan struct {
	projectDir        string
	existingDocs      []string
	existingConfigs   []string
	existingStructure []string
	hasKnowledgeDir   bool
	hasAgentsMD       bool
	hasGit            bool
	hasAltyConfig     bool
	hasMaintenanceDir bool
}

// NewProjectScan creates a ProjectScan value object.
func NewProjectScan(
	projectDir string,
	existingDocs, existingConfigs, existingStructure []string,
	hasKnowledgeDir, hasAgentsMD, hasGit, hasAltyConfig, hasMaintenanceDir bool,
) ProjectScan {
	ed := make([]string, len(existingDocs))
	copy(ed, existingDocs)
	ec := make([]string, len(existingConfigs))
	copy(ec, existingConfigs)
	es := make([]string, len(existingStructure))
	copy(es, existingStructure)
	return ProjectScan{
		projectDir:        projectDir,
		existingDocs:      ed,
		existingConfigs:   ec,
		existingStructure: es,
		hasKnowledgeDir:   hasKnowledgeDir,
		hasAgentsMD:       hasAgentsMD,
		hasGit:            hasGit,
		hasAltyConfig:     hasAltyConfig,
		hasMaintenanceDir: hasMaintenanceDir,
	}
}

// ProjectDir returns the project directory path.
func (s ProjectScan) ProjectDir() string { return s.projectDir }

// HasKnowledgeDir returns whether the project has a knowledge directory.
func (s ProjectScan) HasKnowledgeDir() bool { return s.hasKnowledgeDir }

// HasAgentsMD returns whether the project has an agents.md file.
func (s ProjectScan) HasAgentsMD() bool { return s.hasAgentsMD }

// HasGit returns whether the project has a git repository.
func (s ProjectScan) HasGit() bool { return s.hasGit }

// ExistingDocs returns a defensive copy.
func (s ProjectScan) ExistingDocs() []string {
	out := make([]string, len(s.existingDocs))
	copy(out, s.existingDocs)
	return out
}

// ExistingConfigs returns a defensive copy of existing config file paths.
func (s ProjectScan) ExistingConfigs() []string {
	out := make([]string, len(s.existingConfigs))
	copy(out, s.existingConfigs)
	return out
}

// ExistingStructure returns a defensive copy of existing directory paths.
func (s ProjectScan) ExistingStructure() []string {
	out := make([]string, len(s.existingStructure))
	copy(out, s.existingStructure)
	return out
}

// HasAltyConfig returns whether the project has .alty/config.toml.
func (s ProjectScan) HasAltyConfig() bool { return s.hasAltyConfig }

// HasMaintenanceDir returns whether the project has .alty/maintenance/.
func (s ProjectScan) HasMaintenanceDir() bool { return s.hasMaintenanceDir }

// Gap is a single structural gap found during project analysis.
type Gap struct {
	gapID       string
	gapType     GapType
	path        string
	description string
	severity    GapSeverity
}

// NewGap creates a Gap value object.
func NewGap(gapID string, gapType GapType, path, description string, severity GapSeverity) Gap {
	return Gap{
		gapID:       gapID,
		gapType:     gapType,
		path:        path,
		description: description,
		severity:    severity,
	}
}

// GapID returns the gap identifier.
func (g Gap) GapID() string { return g.gapID }

// GapType returns the gap type classification.
func (g Gap) GapType() GapType { return g.gapType }

// Path returns the path of the missing or conflicting item.
func (g Gap) Path() string { return g.path }

// Description returns a human-readable gap description.
func (g Gap) Description() string { return g.description }

// Severity returns the gap severity.
func (g Gap) Severity() GapSeverity { return g.severity }

// IsDirectory returns true if this gap represents a directory (path ends with "/").
func (g Gap) IsDirectory() bool { return strings.HasSuffix(g.path, "/") }

// MigrationPlan is an immutable plan for resolving gaps.
type MigrationPlan struct {
	planID       string
	branchName   string
	gaps         []Gap
	skipAgentsMD bool
}

// NewMigrationPlan creates a MigrationPlan value object.
func NewMigrationPlan(planID string, gaps []Gap, branchName string, skipAgentsMD bool) MigrationPlan {
	g := make([]Gap, len(gaps))
	copy(g, gaps)
	if branchName == "" {
		branchName = "alty/init"
	}
	return MigrationPlan{
		planID:       planID,
		gaps:         g,
		branchName:   branchName,
		skipAgentsMD: skipAgentsMD,
	}
}

// PlanID returns the plan identifier.
func (p MigrationPlan) PlanID() string { return p.planID }

// BranchName returns the git branch name for the migration.
func (p MigrationPlan) BranchName() string { return p.branchName }

// SkipAgentsMD returns whether to skip generating agents.md.
func (p MigrationPlan) SkipAgentsMD() bool { return p.skipAgentsMD }

// Gaps returns a defensive copy.
func (p MigrationPlan) Gaps() []Gap {
	out := make([]Gap, len(p.gaps))
	copy(out, p.gaps)
	return out
}

// GapAnalysis is the aggregate root for the rescue flow.
type GapAnalysis struct {
	analysisID    string
	projectDir    string
	status        AnalysisStatus
	scan          *ProjectScan
	gaps          []Gap
	plan          *MigrationPlan
	failureReason string
	events        []GapAnalysisCompleted
}

// NewGapAnalysis creates a new GapAnalysis aggregate root.
func NewGapAnalysis(projectDir string) *GapAnalysis {
	return &GapAnalysis{
		analysisID: identity.NewID(),
		projectDir: projectDir,
		status:     AnalysisStatusScanning,
	}
}

// AnalysisID returns the analysis identifier.
func (a *GapAnalysis) AnalysisID() string { return a.analysisID }

// ProjectDir returns the project directory.
func (a *GapAnalysis) ProjectDir() string { return a.projectDir }

// Status returns the current analysis status.
func (a *GapAnalysis) Status() AnalysisStatus { return a.status }

// Scan returns the project scan result, or nil.
func (a *GapAnalysis) Scan() *ProjectScan { return a.scan }

// Gaps returns a defensive copy.
func (a *GapAnalysis) Gaps() []Gap {
	out := make([]Gap, len(a.gaps))
	copy(out, a.gaps)
	return out
}

// Plan returns the migration plan, or nil.
func (a *GapAnalysis) Plan() *MigrationPlan { return a.plan }

// Events returns a defensive copy of domain events.
func (a *GapAnalysis) Events() []GapAnalysisCompleted {
	out := make([]GapAnalysisCompleted, len(a.events))
	copy(out, a.events)
	return out
}

// SetScan records scan results. Only from SCANNING state.
func (a *GapAnalysis) SetScan(scan ProjectScan) error {
	if a.status != AnalysisStatusScanning {
		return fmt.Errorf("cannot set scan in %s state: %w", string(a.status), domainerrors.ErrInvariantViolation)
	}
	a.scan = &scan
	return nil
}

// Analyze sets analysis results. Requires scan to exist.
func (a *GapAnalysis) Analyze(gaps []Gap) error {
	if a.status != AnalysisStatusScanning {
		return fmt.Errorf("cannot analyze in %s state: %w", string(a.status), domainerrors.ErrInvariantViolation)
	}
	if a.scan == nil {
		return fmt.Errorf("cannot analyze without scan: %w", domainerrors.ErrInvariantViolation)
	}
	g := make([]Gap, len(gaps))
	copy(g, gaps)
	a.gaps = g
	a.status = AnalysisStatusAnalyzed
	return nil
}

// CreatePlan creates a migration plan. Only from ANALYZED.
func (a *GapAnalysis) CreatePlan(plan MigrationPlan) error {
	if a.status != AnalysisStatusAnalyzed {
		return fmt.Errorf("cannot create plan in %s state: %w", string(a.status), domainerrors.ErrInvariantViolation)
	}
	a.plan = &plan
	a.status = AnalysisStatusPlanned
	return nil
}

// BeginExecution starts executing the plan. Only from PLANNED.
func (a *GapAnalysis) BeginExecution() error {
	if a.status != AnalysisStatusPlanned {
		return fmt.Errorf("cannot begin execution in %s state: %w", string(a.status), domainerrors.ErrInvariantViolation)
	}
	a.status = AnalysisStatusExecuting
	return nil
}

// Complete marks as completed and emits event.
func (a *GapAnalysis) Complete() error {
	if a.status != AnalysisStatusExecuting {
		return fmt.Errorf("cannot complete in %s state: %w", string(a.status), domainerrors.ErrInvariantViolation)
	}
	a.status = AnalysisStatusCompleted
	a.events = append(a.events, NewGapAnalysisCompleted(
		a.analysisID, a.projectDir, len(a.gaps), len(a.gaps),
	))
	return nil
}

// Fail marks as failed and records the reason. Only from EXECUTING.
func (a *GapAnalysis) Fail(reason string) error {
	if a.status != AnalysisStatusExecuting {
		return fmt.Errorf("cannot fail in %s state: %w", string(a.status), domainerrors.ErrInvariantViolation)
	}
	a.status = AnalysisStatusFailed
	a.failureReason = reason
	return nil
}

// FailureReason returns the reason for failure, or empty string if not failed.
func (a *GapAnalysis) FailureReason() string { return a.failureReason }
